src/coordinator/ts/values.go

```diff
@@ -39,79 +38,98 @@ type Values interface {
-	values := make([]float64, numSteps)
+type fixedResolutionValues struct {
+	millisPerStep time.Duration
+	numSteps      int
+	values        []float64
+	startTime     time.Time
+}
 
-	// Faster way to initialize an array instead of a loop
-	Memset(values, initialValue)
-	vals := &float64Values{
-		ctx:           ctx,
-		millisPerStep: millisPerStep,
-		numSteps:      numSteps,
-		allNaN:        math.IsNaN(initialValue),
-		values:        values,
+func (b *fixedResolutionValues) MillisPerStep() time.Duration { return b.millisPerStep }
+func (b *fixedResolutionValues) Len() int                     { return b.numSteps }
+func (b *fixedResolutionValues) ValueAt(point int) float64    { return b.values[point] }
+func (b *fixedResolutionValues) DatapointAt(point int) Datapoint {
+	return Datapoint{
+		Timestamp: b.StartTimeForStep(point),
+		Value:     b.ValueAt(point),
 	}
+}
 
-	return vals
+// StartTime returns the time the values start
+func (b *fixedResolutionValues) StartTime() time.Time {
+	return b.startTime
 }
 
-type float64Values struct {
-	ctx           context.Context
-	millisPerStep int
-	numSteps      int
-	values        []float64
-	allNaN        bool
+// Resolution returns resolution per step
+func (b *fixedResolutionValues) Resolution() time.Duration {
+	return b.MillisPerStep()
 }
 
-func (b *float64Values) Reset() {
-	for i := range b.values {
-		b.values[i] = math.NaN()
-	}
-	b.allNaN = true
+// StepAtTime returns the step within the block containing the given time
+func (b *fixedResolutionValues) StepAtTime(t time.Time) int {
+	return int(t.Sub(b.StartTime()) / b.MillisPerStep())
+}
+
+// StartTimeForStep returns the time at which the given step starts
+func (b *fixedResolutionValues) StartTimeForStep(n int) time.Time {
+	return b.startTime.Add(time.Duration(n) * b.MillisPerStep())
 }
 
-func (b *float64Values) AllNaN() bool              { return b.allNaN }
-func (b *float64Values) MillisPerStep() int        { return b.millisPerStep }
-func (b *float64Values) Len() int                  { return b.numSteps }
-func (b *float64Values) ValueAt(point int) float64 { return b.values[point] }
-func (b *float64Values) SetValueAt(point int, v float64) {
-	b.allNaN = b.allNaN && math.IsNaN(v)
-	b.values[point] = v
+// SetValueAt sets the value at the given entry
+func (b *fixedResolutionValues) SetValueAt(n int, v float64) {
+	b.values[n] = v
 }
 
-func (b *float64Values) Slice(begin, end int) Values {
-	return &float64Values{
-		ctx:           b.ctx,
-		millisPerStep: b.millisPerStep,
-		values:        b.values[begin:end],
-		numSteps:      end - begin,
-		allNaN:        false,
+// NewFixedStepValues returns mutable values with fixed resolution
+func NewFixedStepValues(millisPerStep time.Duration, numSteps int, initialValue float64, startTime time.Time) FixedResolutionMutableValues {
+	values := make([]float64, numSteps)
+	// Faster way to initialize an array instead of a loop
+	Memset(values, initialValue)
```

Comment:
Do we need to initialize a value at all?

Reply:
we'd be pooling this (didn't want to do that in this diff), so we would surely need initialization